### ISSUE TITLE
PyTorch models need to be in eval mode in order to work

### ISSUE THREAD

**Issue by TDeVries** [bethgelab/foolbox#74]
I spent a while trying to get foolbox to produce adversarial examples using a pytorch model, but it mostly returned adversaries that were still no problem for the network to solve. The solution was simply to put the model in eval mode instead of train mode, since otherwise dropout and batch norm will constantly change the network and make finding an adversary quite difficult (somewhat obvious in hindsight). 

Would it be possible to do a check to make sure the model is in eval mode before trying to create adversarial examples, or at least make a note of it somewhere in the docs so that other people don't run into the same issue?

**Comment by jonasrauber** [comment]
@TDeVries Is there a standard way in PyTorch to check if a model is in eval mode?

**Comment by TDeVries** [comment]
You should be able to check the training state of the model:

```
if model.training == True:
    # Train mode
if model.training == False:
    # Evaluation mode
```

The .eval() function:
https://github.com/pytorch/pytorch/blob/a64daf2c5975bfae53c43ede88c3a84aa4eadad7/torch/nn/modules/module.py#L538

And right above it is the .train() function, which is called by .eval and sets the model's training state:
https://github.com/pytorch/pytorch/blob/a64daf2c5975bfae53c43ede88c3a84aa4eadad7/torch/nn/modules/module.py#L528

**Comment by jonasrauber** [comment]
I created a pull-request that will produce a warning if the model is in train mode. There can be legitimate reasons to attack a model in train mode, but I agree that it's usually not what one wants and therefore a warning makes sense.

**Comment by muzi-8** [comment]
@TDeVries 
Hi, I also want to produce adversarial examples using a pytorch model, should you provide a template example by pytorch framework ?
Thanks!

**Comment by jonasrauber** [comment]
@muzi-8 You can have a look at the tests: https://github.com/bethgelab/foolbox/blob/v1.1.0/foolbox/tests/test_models_pytorch.py#L10

But you are right: it would be nice to have more examples in the documentation. Feel free to make a pull-request.

**Comment by muzi-8** [comment]
@jonasrauber @TDeVries 
Sorry,I'm a novice. I would cppreciate it if you help me.Thanks!

**Comment by jonasrauber** [comment]
Please have a look at the docs http://foolbox.readthedocs.io/en/latest/
